### DISCUSSION TITLE
Remove Bounce Rate and replace it with Engaged Visitors

### DISCUSSION THREAD

**Discussion by metmarkosaric** [plausible/analytics#591] [Anything Else]
Bounce rate has always been a bit of a weird metric in the world of web analytics. 

In the past we've spoken about replacing it with something inverse of bounce rate that rolls multiple metrics into a single engagement score. The main limiting factor was that it's not the industry standard so would need a lot of communication and explanation. But now in the GA4 version they're removing Bounce Rate and replacing it with Engaged Session which makes it easier for us to explain a similar move.

For a visitor to qualify as Engaged, the visitor must do at least one of the following or something along similar lines during their visit:

* Visit 2 or more pages
* Fire a conversion event
* Actively engage with the site for at least X seconds
* Any other ideas?

Any thoughts on this?

**Comment by Vigasaurus** [comment]
I think those make a lot of sense, one thing to consider in my opinion though is that some people (myself included) use a conversion event as a kind of pulse event, that fires on an entry to track various metrics through the event props. For myself it includes various generalized user settings and similar (think font size or theme-selection for a blog example). I may well be in the minority, but I think if a conversion event is able to call a user "engaged", having a way to call specific conversions as "non-engaging" would be useful. Maybe something like `plausible('custom', {engaging: false, props: {...}})`, with it defaulting to true, of course. 

I do also think that it could be very neat, if this method was employed, to have access to a filter of `Engaged` sessions, but I recognize that wouldn't be very backwards compatible and I personally don't remember the session structure well enough to speak on this idea's complexity - but just an idea.

**Reply by metmarkosaric** [reply]
thanks @Vigasaurus! makes sense. GA allows you to set nonInteraction which does pretty much what you're describing so that's necessary to have for use cases like yours.

**Reply by ukutaht** [reply]
> I think those make a lot of sense, one thing to consider in my opinion though is that some people (myself included) use a conversion event as a kind of pulse event, that fires on an entry to track various metrics through the event props. For myself it includes various generalized user settings and similar (think font size or theme-selection for a blog example). I may well be in the minority, but I think if a conversion event is able to call a user "engaged", having a way to call specific conversions as "non-engaging" would be useful. Maybe something like `plausible('custom', {engaging: false, props: {...}})`, with it defaulting to true, of course.
> 
> I do also think that it could be very neat, if this method was employed, to have access to a filter of `Engaged` sessions, but I recognize that wouldn't be very backwards compatible and I personally don't remember the session structure well enough to speak on this idea's complexity - but just an idea.

This is a good workaround for now but ideally capturing this sort of information would be built into Plausible so you don't need to fire a custom event for tracking say `theme-selection`. See https://github.com/plausible/analytics/discussions/320#discussioncomment-179904 for more discussion on how to avoid these sorts of extra calls.

In any case, I also agree with adding support for non-engagement events. Currently we treat every event as 'engagement' event which skews bounce rate when you use custom events for extra session/visitor tracking as you are.

**Reply by Vigasaurus** [reply]
I just wanted to give this a bump (only the non-engagement events part)/get confirmation if the API that I'd mentioned briefly in the parent comment here is what we'd want (I can PR eventually, if so) - only reason being is that I got reminded recently of when I skewed my bounce rate metrics really badly at one point when I added a custom event on one of my main landing pages for a very minimal interaction (just a hover on a new interactive element) after graphing bounce rate, heh. 

Maybe with the API being an exclusion rather than an inclusion, akin to GA - i.e.
```
plausible('custom', {nonInteraction: true, props: {...}})
```

![image](https://user-images.githubusercontent.com/19434157/136959033-f44666d7-cf50-46d5-8ae1-f160bb9ced76.png)

**Reply by ukutaht** [reply]
I'm on the fence with this one. I'm curious to get some feedback since this decision will probably also apply to any future options like this.

On one hand, the API you suggested is what I would want as the user of it. The `nonInteraction` key is obviously not part of the custom props I'm sending so it should be kept separate.

On the other hand, we could treat `nonInteraction` as a _special_ or _reserved_ property of an event. The API would then be:
```js
plausible('custom', {props: {nonInteraction: true, ...}})
```

The benefit of doing this is that we don't need to add any special code in the script to handle this option and the script would stay smaller. It could extend to future options as well. Although I can't think of many special properties other than `currency` for ecommerce and overriding `userId`.

The drawback is that:
1. The API muddles the concepts of custom props and special event props into a single thing
2. We have to preserve the keys as they are specified e.g. `nonInteraction` in the JSON payload. If we had a special handler for it in the script, it could compress they key to `ni: true` like GA does and save bandwidth that way.

I don't know, maybe I'm overthinking it. I'm not kidding when I say we obsess over the payload size!

**Reply by ukutaht** [reply]
Another comment, and it might just be me but to me writing `interaction: false` would be more natural than `nonInteraction: true`. What do you think?

**Reply by Vigasaurus** [reply]
Yeah that's a good point about the script size - maybe there's some smart stuff we can do in there to just loop over the valid options (props & meta & now interactivity) and stringify them dynamically instead of doing them all separately, it probably just needs some testing, but I do think some optimization and it could be made to match the existing setup in size (or at least be very close). 

As far as it being `interaction: false`, I did originally think that too, but it feels kinda weird for the default for a property like that being treated as truthy when not defined, no? I'm probably also overthinking it though, it does feel more natural just a bit ... *wrong* as far as how truthiness works heh

**Reply by ukutaht** [reply]
Eh, it just defaults to `true` when not defined. I don't see anything wrong with that :smile: 

I would really like to remove the `meta` field. I don't think anyone uses it but I have to verify before removing it.

Thinking about it a bit more I'm definitely leaning towards putting this in `props` at the moment. It's also consistent with how `pageview` is a special event name instead of having some sort of separate 'type' attribute on an event or something.

**Reply by Vigasaurus** [reply]
Actually a note here - to both save script size, and make it extremely extensible, wouldn't it just make most sense for the entire `options` variable from `trigger(eventName, options)` to get stringified and passed along? Then we can offload all the parsing etc to elixir, while keeping the script even smaller than it is now (since it has the duplication for props & meta, and that could become just one stringify with this solution)

**Reply by ukutaht** [reply]
Yeah I had the same though but then I realized we let people send `callback` as an option as well. That should definitely not be serialized and sent.

**Reply by Vigasaurus** [reply]
Huh I somehow never noticed that was even an option - yeah I guess that does complicate things a bit though, especially since any extra changes there (e.g. moving the callback out to be the third param) would be a backwards compatibility/documentation nightmare.

**Comment by oliverpool** [comment]
> Actively engage with the site for at least X seconds

Since the `X` could depend on the page (did the user read my whole blog post before quitting, or were they bored after the introduction?), I think it would be very useful to still know how long a page-visit lasted (see my previous comment regarding a possible implementation https://github.com/plausible/analytics/discussions/111#discussioncomment-179766)

**Reply by metmarkosaric** [reply]
Yes, time on individual pages is still the main metric I personally miss in Plausible and we'll for sure add it at some point.

**Comment by ukutaht** [comment]
> Actively engage with the site for at least X seconds

I wonder how this is implemented in GA. There are so many ways to skin this cat that any standardization on this metric is probably impossible.

**Reply by luciandavidescu** [reply]
"(...) during their visit:"
- Scroll until the end of the main content

**Comment by Dan0sz** [comment]
Maybe clicking on an outgoing link should count as engaged as well? It does show that the visitor read the content.

**Reply by metmarkosaric** [reply]
yeah any of the different goals/events could be included (file download, external link click...)

**Reply by ukutaht** [reply]
I suppose it would count, sort of, as a 'conversion event'?
